### PR TITLE
Fix reversed output in interpolation concatenation

### LINKED ISSUE #55461: Binary addition of interpolated strings mangles the string
**Version Used**: 

.NET `6.0.100-rc.1.21405.1` (`6.0.100-rc.1.21403.66` works fine)

**Steps to Reproduce**:

Test code:

```csharp
using System;

class App {
	public static void Main (string[] args)
	{
		var t = new T ();
		Console.WriteLine (t.A ());
		Console.WriteLine ();
		Console.WriteLine (t.B ());
	}
}

class T {
	public float R0C0 = 1;
	public float R1C0 = 2;
	public float R0C1 = 3;
	public float R1C1 = 4;

	public string A ()
	{
		return
			$"({R0C0}, {R0C1})\n" +
			$"({R1C0}, {R1C1})";
	}

	public string B ()
	{
		return
@$"({R0C0}, {R0C1})
({R1C0}, {R1C1})";
	}
}
```

csproj:

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net6.0</TargetFramework>
    <LangVersion>latest</LangVersion>
  </PropertyGroup>
</Project>
```

Save & run with `dotnet run`

**Expected Behavior**:

Output should be:

```
(1, 3)
(2, 4)

(1, 3)
(2, 4)

```

The first interpolated string is badly mangled.

**Actual Behavior**:

```
)
3, 1()4, 2(

(1, 3)
(2, 4)
```

This is a regression in this range: https://github.com/dotnet/roslyn/compare/3a06c12480787c39e37354dd5d171fa6d91dece1...38582c4d3f32679e5cc29bf86c4bde012ed2baee

From the PR description, this seems like a likely culprit: https://github.com/dotnet/roslyn/pull/55059

### LINKED ISSUE #55493: Interpolated string regression?
I've noticed some ASP.NET Core tests failing to run locally, and upon inspection with the debugger seems that something is going on with string interpolation.

When I concatenate the string as follows I get the proper text:
```csharp
var expectedMessage = "The type '" + typeof(HasParameterClashingWithInherited).FullName + "' declares more than one parameter matching the " +
                "name '"+ nameof(HasParameterClashingWithInherited.IntProp).ToLowerInvariant() + "'. Parameter names are case-insensitive and must be unique.";
```
```
The type 'Microsoft.AspNetCore.Components.ParameterViewTest+HasParameterClashingWithInherited' declares more than one parameter matching the name 'intprop'. Parameter names are case-insensitive and must be unique.
```

However, when I do the same thing via string interpolation the results are completely different.
```csharp
$"The type '{typeof(HasParameterClashingWithInherited).FullName}' declares more than one parameter matching the " +
                $"name '{nameof(HasParameterClashingWithInherited.IntProp).ToLowerInvariant()}'. Parameter names are case-insensitive and must be unique."
```
This is what I get upon putting the code into a variable and inspecting it. Seems like all the placeholders are empty.
```
' declares more than one parameter matching the Microsoft.AspNetCore.Components.ParameterViewTest+HasParameterClashingWithInheritedThe type ''. Parameter names are case-insensitive and must be unique.intpropname '
```

Note that this is not a test I changed in any way as part of my changes, so I would just expect it to run. The fact that avoiding the interpolated string worksaround the issue seems to point out that there's some compilation/runtime combo going on here.

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #55493
Fixes #55461

Relates to feature/test plan: https://github.com/dotnet/roslyn/issues/51499

### CODE REVIEW

**Review by 333fred — CHANGES_REQUESTED** [review]
Don't use a reverse on the parts arrays, reverse the builder after populating.

**Comment by Youssef1313** [comment]
@333fred The builder is indeed reversed at the end of the method, but that wasn't enough. To reverse a builder, I'd have to create a  new builder for each `BoundInterpolatedString`, or loop backwards on `Parts` and use `Add` for each part (instead of `AddRange(Parts)`.

Alternatively, attempt to rewrite the code such that we don't need reversing at all (ie, walk from left to right - but I'm not sure how hard is that or if that's possible).

**Comment by 333fred** [comment]
Left to right isn't feasible. If we need to add these in reverse order here as well, we should not use linq to do so.

**Comment by 333fred** [comment]
I'm also concerned about flow analysis. Make sure we have a test like `$"{a = null}{a.ToString()}" + $""`

File: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs [hunk 1]
```diff
@@ -12182,6 +12182,93 @@ .maxstack  3
...
+");
+        }
+
+        [Fact]
+        public void InterpolatedStringsAddedUnderObjectAddition3()
+        {
+            var code = @"
+using System;
+
+try
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:12218]: Needs to be nullable enabled to test the flow analysis.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs [hunk 2]
```diff
@@ -5,6 +5,7 @@
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Diagnostics;
+using System.Linq;
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs:8]: Unused.

**Review by 333fred — APPROVED** [review]
@jcouv ptal.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 4)

**Comment by jcouv** [comment]
Looks like there is a legitimate failure: `InterpolatedStringsAddedUnderObjectAddition3` in the spanish CI branch because of localization of `"Object reference not set to an instance of an object."`. Consider just printing the exception type name.

**Comment by jcouv** [comment]
Thanks @Youssef1313

**Pull request merged**